samples/nrf9160/serial_lte_modem/src/slm_at_tcpip.c

```diff
@@ -0,0 +1,716 @@
+	int ret;
+	char buf[16];
+
+	ret = do_udp_init(url, port);
+	CHECK_ERROR(ret);
+
+	while (offset < datalen) {
+		ret = sendto(client.sock, data + offset,
+			   datalen - offset, 0,
+			   (struct sockaddr *)&remote,
+			   sizeof(struct sockaddr_in));
+		if (ret <= 0) {
+			LOG_ERR("sendto() failed: %d", -errno);
+			do_socket_close(-errno);
+			return -errno;
+		}
+
+		offset += ret;
+	}
+
+	sprintf(buf, "#XUDPSENDTO: %d\r\n", offset);
+	client.callback(buf);
+	LOG_DBG("UDP sent");
+	return 0;
+}
+
+static int do_udp_recvfrom(const char *url, u16_t port, u16_t length,
+			u16_t time)
+{
+	int ret;
+	char data[NET_IPV4_MTU];
+	int sockaddr_len = sizeof(struct sockaddr);
+	char buf[16];
+	struct timeval tmo = {
+		.tv_sec = time
+	};
+
+	ret = do_udp_init(url, port);
+	CHECK_ERROR(ret);
+
+	ret = setsockopt(client.sock, SOL_SOCKET, SO_RCVTIMEO,
+			&tmo, sizeof(struct timeval));
+	if (ret < 0) {
+		LOG_ERR("setsockopt() error: %d", -errno);
+		do_socket_close(-errno);
+		return ret;
+	}
+
+	if (length > NET_IPV4_MTU) {
+		ret = recvfrom(client.sock, data, NET_IPV4_MTU, 0,
+			(struct sockaddr *)&remote, &sockaddr_len);
+	} else {
+		ret = recvfrom(client.sock, data, length, 0,
+			(struct sockaddr *)&remote, &sockaddr_len);
+	}
+	if (ret < 0) {
+		LOG_WRN("recvfrom() error: %d", -errno);
+		do_socket_close(-errno);
+		ret = -errno;
+	} else {
+		/*
+		 * Datagram sockets in various domains permit zero-length
+		 * datagrams. When such a datagram is received, the return
+		 * value is 0. Treat as normal case
+		 */
+		data[ret] = 0x00;
+		client.callback("#XUDPRECV: ");
+		client.callback(data);
+		client.callback("\r\n");
+		sprintf(buf, "#XUDPRECV: %d\r\n", ret);
+		client.callback(buf);
+		ret = 0;
+	}
+
+	LOG_DBG("UDP received");
+	return ret;
+}
+
+/**@brief handle AT#XSOCKET commands
+   AT#XSOCKET=<op>[,<type>]
+   AT#XSOCKET?
+   AT#XSOCKET=? TEST command not supported
+ */
+static int handle_at_socket(const char *at_cmd_param)
+{
+	int err = -EINVAL;
+	char* at_param = (char *)at_cmd_param;
+	u16_t op;
+
+	if (*(at_param) == '=') {
+		at_param++;
+		if (*(at_param) == '?') {
+			return err;
+		}
+		err = at_parser_params_from_str(at_param, NULL, &m_param_list);
+		CHECK_ERROR(err);
+		if (at_params_valid_count_get(&m_param_list) < 1) {
+			return -EINVAL;
+		}
+		err = at_params_short_get(&m_param_list, 0, &op);
+		CHECK_ERROR(err);
+		if (op == 1) {
+			u16_t type;
+			if (at_params_valid_count_get(&m_param_list) < 2) {
+				return -EINVAL;
+			}
+			err = at_params_short_get(&m_param_list, 1, &type);
+			CHECK_ERROR(err);
+			if (client.sock > 0) {
+				LOG_WRN("Socket is already opened");
+			} else {
+				err = do_socket_open(type);
+			}
+		} else if (op == 0) {
+			if (client.sock < 0) {
+				LOG_WRN("Socket is not opened yet");
+			} else {
+				err = do_socket_close(0);
+			}
+		}
+	} else if (*(at_param) == '?') {
+		char buf[16];
+		if (client.sock != INVALID_SOCKET) {
+			sprintf(buf, "#XSOCKET: %d, %d\r\n", client.sock,
```

Comment:
Also here, more than 16 bytes.

Reply:
Like the gps part, use a global 64-byte buffer now.